src/main/host/managed_thread.rs

```diff
@@ -0,0 +1,611 @@
+            Ok(ShimEvent::SyscallComplete(res)) => res.retval,
+            Err(SelfContainedChannelError::WriterIsClosed) => {
+                trace!("Plugin exited while executing native syscall {n}");
+                process.mark_as_exiting();
+                self.cleanup();
+                SysCallReg::from(-(Errno::ESRCH as i64))
+            }
+            other => {
+                panic!("Unexpected response from plugin: {other:?}");
+            }
+        }
+    }
+
+    pub fn run(
+        &mut self,
+        host: &Host,
+        plugin_path: &CStr,
+        argv: Vec<CString>,
+        mut envv: Vec<CString>,
+        working_dir: &CStr,
+        strace_fd: Option<RawFd>,
+        log_path: &CStr,
+    ) {
+        envv.push(
+            CString::new(format!(
+                "SHADOW_IPC_BLK={}",
+                self.ipc_shmem.serialize().encode_to_string()
+            ))
+            .unwrap(),
+        );
+        envv.push(CString::new(format!("SHADOW_PID={}", getpid().as_raw())).unwrap());
+        envv.push(CString::new(format!("SHADOW_TSC_HZ={}", host.tsc().cyclesPerSecond)).unwrap());
+
+        info!("forking new mthread with environment '{envv:?}', arguments '{argv:?}', and working directory '{working_dir:?}'");
+
+        let shimlog_fd = nix::fcntl::open(
+            log_path,
+            OFlag::O_WRONLY | OFlag::O_CREAT | OFlag::O_CLOEXEC,
+            Mode::S_IRUSR | Mode::S_IWUSR | Mode::S_IRGRP | Mode::S_IROTH | Mode::S_IWOTH,
+        )
+        .unwrap();
+
+        let child_pid = self.spawn(plugin_path, argv, envv, working_dir, strace_fd, shimlog_fd);
+
+        // should be opened in the shim, so no need for it anymore
+        nix::unistd::close(shimlog_fd).unwrap();
+
+        // In Linux, the PID is equal to the TID of its first thread.
+        self.native_pid = Some(child_pid);
+        self.native_tid = Some(child_pid);
+
+        // Configure the child_pid_watcher to close the IPC channel when the child dies.
+        let handle = {
+            let ipc = self.ipc_shmem.clone();
+            WORKER_SHARED
+                .borrow()
+                .as_ref()
+                .unwrap()
+                .child_pid_watcher()
+                .register_callback(child_pid, move |_pid| {
+                    ipc.from_plugin().close_writer();
+                })
+        };
+        self.notification_handle = Some(handle);
+
+        // When `continue`d we'll tell the plugin to start executing.
+        self.current_event = ShimEvent::Start;
+
+        self.is_running = true;
+    }
+
+    pub fn resume(
+        &mut self,
+        thread: &Thread,
+        process: &Process,
+        host: &Host,
+    ) -> Option<SysCallCondition> {
+        debug_assert!(self.is_running());
+
+        self.sync_affinity_with_worker();
+
+        // Flush any pending writes, e.g. from a previous mthread that exited
+        // without flushing.
+        process.free_unsafe_borrows_flush().unwrap();
```

Comment:
Maybe add a comment that this only is needed for C syscall handlers (I think?)?

Reply:
Right; done